File: android-client/lib/services/friend_service.dart
import 'package:android_client/classes/friend_notifier.dart';
import 'package:android_client/classes/user.dart';
import 'package:android_client/services/socket_service.dart';
import 'package:android_client/services/user_http_service.dart';
import 'package:android_client/services/user_service.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

enum UserState {
  isFriend,
  requestReceived,
  requestUnseen,
  requestSent,
  isStranger,
}

enum FriendEvent {
  seenOneRequest,
  seenRequests,
  registerId,
  removeId,
  request,
  sentRequest,
  receiveRequest,
  acceptRequest,
  declineRequest,
  removeFriend,
  addNewUser
}

class FriendService {
  late final FriendNotifier _friendNotifier;
  final SocketService _socketService = SocketService();
  final UserService _userService = UserService();

  FriendService();

  void setUsers(Function setUsersCallback) {
    _userService
        .getUserList()
        .then((List<UserProfile> users) => {setUsersCallback(users)});
  }

  void registerFriendSocket() {
    _socketService.send(
        FriendEvent.registerId.name, UserService.loggedInUser?.id);
  }

  void deregisterFriendSocket() {
    _socketService.send(
        FriendEvent.removeId.name, UserService.loggedInUser?.id);
  }

  void seenFriendRequests() {
    _socketService.send(
        FriendEvent.seenRequests.name, UserService.loggedInUser?.id);
  }

  void sendFriendRequest(String userId) {
    _socketService.send(FriendEvent.request.name,
        <String, String?>{'from': UserService.loggedInUser?.id, 'to': userId});
  }

  void acceptFriendRequest(String userId) {
    _socketService.send(FriendEvent.acceptRequest.name,
        <String, String?>{'from': userId, 'to': UserService.loggedInUser?.id});
  }

  void declineFriendRequest(String userId) {
    _socketService.send(FriendEvent.declineRequest.name,
        <String, String?>{'from': userId, 'to': UserService.loggedInUser?.id});
  }

  void removeFriend(String userId) {
    _socketService.send(FriendEvent.removeFriend.name,
        <String, String?>{'from': UserService.loggedInUser?.id, 'to': userId});
  }

  void resetSocketEvents(BuildContext context) {
    _friendNotifier = Provider.of<FriendNotifier>(context, listen: false);

    unbindSocketEvents();

    _socketService.on(FriendEvent.sentRequest.name, (dynamic userId) async {
      _friendNotifier.propagateEvent(FriendEvent.sentRequest.name, userId);
    });

    _socketService.on(FriendEvent.request.name, (dynamic userId) async {
      _friendNotifier.propagateEvent(FriendEvent.request.name, userId);
    });

    _socketService.on(FriendEvent.acceptRequest.name, (dynamic userId) async {
      _friendNotifier.propagateEvent(FriendEvent.acceptRequest.name, userId);
    });

    _socketService.on(FriendEvent.declineRequest.name, (dynamic userId) async {
      _friendNotifier.propagateEvent(FriendEvent.declineRequest.name, userId);
    });

    _socketService.on(FriendEvent.removeFriend.name, (dynamic userId) async {
      _friendNotifier.propagateEvent(FriendEvent.removeFriend.name, userId);
    });

    _socketService.on(FriendEvent.seenOneRequest.name, (dynamic userId) async {
      _friendNotifier.propagateEvent(FriendEvent.seenOneRequest.name, '');
    });

    _socketService.on(FriendEvent.seenRequests.name, (dynamic userId) async {
      _friendNotifier.propagateEvent(FriendEvent.seenRequests.name, '');
    });

    _socketService.on(FriendEvent.addNewUser.name, (dynamic user) async {
      user = await UserProfile.fromJson(user);
      UserHttpService.getReplacedUrl(user.avatar);
      _friendNotifier.addNewUser(FriendEvent.addNewUser.name, user);
    });
  }

  void unbindSocketEvents() {
    _socketService.off(FriendEvent.sentRequest.name);
    _socketService.off(FriendEvent.request.name);
    _socketService.off(FriendEvent.acceptRequest.name);
    _socketService.off(FriendEvent.declineRequest.name);
    _socketService.off(FriendEvent.removeFriend.name);
    _socketService.off(FriendEvent.seenOneRequest.name);
    _socketService.off(FriendEvent.seenRequests.name);
    _socketService.off(FriendEvent.addNewUser.name);
  }
}

File: windows-client/src/app/services/friend/friend.service.ts
/* eslint-disable no-underscore-dangle */
import { Injectable } from '@angular/core';
import { CommunicationSocketService } from '@app/services/communication-socket/communication-socket.service';
import { UserService } from '@app/services/user/user.service';
import { UserProfile } from '@common/classes/user-profile';
import { FriendEvents } from '@common/enums/friend.gateway.variables';
import { Subject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class FriendService {
    requestSent$: Subject<string> = new Subject<string>();
    requestReceived$: Subject<string> = new Subject<string>();
    requestAccepted$: Subject<string> = new Subject<string>();
    requestDeclined$: Subject<string> = new Subject<string>();
    requestsSeen$: Subject<null> = new Subject<null>();
    requestSeenOne$: Subject<null> = new Subject<null>();
    removeFriend$: Subject<string> = new Subject<string>();
    addUser$: Subject<UserProfile> = new Subject<UserProfile>();

    userId: string = '';

    constructor(private readonly userService: UserService, private readonly socketService: CommunicationSocketService) {}

    sendFriendRequest(userId: string) {
        this.socketService.send(FriendEvents.Request, { from: this.userService.loggedInUser?._id as string, to: userId });
    }

    acceptFriendRequest(userId: string) {
        this.socketService.send(FriendEvents.AcceptRequest, { from: userId, to: this.userService.loggedInUser?._id as string });
    }

    declineFriendRequest(userId: string) {
        this.socketService.send(FriendEvents.DeclineRequest, { from: userId, to: this.userService.loggedInUser?._id as string });
    }

    removeFriend(userId: string) {
        this.socketService.send(FriendEvents.RemoveFriend, { from: this.userService.loggedInUser?._id as string, to: userId });
    }

    seenFriendRequests() {
        this.socketService.send(FriendEvents.SeenRequests, this.userService.loggedInUser?._id);
    }

    registerFriendSocket() {
        this.socketService.send(FriendEvents.RegisterId, this.userService.loggedInUser?._id as string);
    }

    deregisterFriendSocket() {
        this.socketService.send(FriendEvents.RemoveId, this.userService.loggedInUser?._id as string);
    }

    registerSocketEvents(): void {
        this.socketService.on(FriendEvents.SentRequest, (userId: string) => {
            this.requestSent$.next(userId);
        });

        this.socketService.on(FriendEvents.Request, (userId: string) => {
            this.requestReceived$.next(userId);
        });

        this.socketService.on(FriendEvents.AcceptRequest, (userId: string) => {
            this.requestAccepted$.next(userId);
        });

        this.socketService.on(FriendEvents.DeclineRequest, (userId: string) => {
            this.requestDeclined$.next(userId);
        });

        this.socketService.on(FriendEvents.RemoveFriend, (userId: string) => {
            this.removeFriend$.next(userId);
        });

        this.socketService.on(FriendEvents.SeenRequests, () => {
            this.requestsSeen$.next(null);
        });

        this.socketService.on(FriendEvents.AddNewUser, (user: UserProfile) => {
            this.addUser$.next(user);
        });

        this.socketService.on(FriendEvents.SeenOneRequest, () => {
            this.requestSeenOne$.next(null);
        });
    }

    deregisterSocketEvents(): void {
        this.socketService.off(FriendEvents.SentRequest);
        this.socketService.off(FriendEvents.Request);
        this.socketService.off(FriendEvents.AcceptRequest);
        this.socketService.off(FriendEvents.DeclineRequest);
        this.socketService.off(FriendEvents.RemoveFriend);
        this.socketService.off(FriendEvents.SeenRequests);
        this.socketService.off(FriendEvents.SeenOneRequest);
        this.socketService.off(FriendEvents.AddNewUser);
    }
}
